### ISSUE TITLE
从Tensor到Transformer——构建 GPT 模型 | Task 4 优化器与训练循环

### ISSUE THREAD

**Issue by lynnyulinlin-debug** [lynnyulinlin-debug/mlsys_on_device#23]
### 1. 目标
   实现优化器和完整的训练循环，完成“前向-损失-反向-更新”的闭环。

### 2. 学习资料

    2.1 07 Optimizers  https://mlsysbook.ai/tinytorch/modules/07_optimizers.html

    2.2 08 Training  https://mlsysbook.ai/tinytorch/modules/08_training.html

### 3.任务

    3.1  任务1（对应2.1）：实现 SGD（含 momentum）和 Adam 优化器

    3.2  任务2（对应2.2）：实现完整的 train 循环，整合 DataLoader、模型、损失函数、优化器
     
### 4.产出

    4.1  Optimizers实现思路, 测试截图

    4.2 Training实现思路, 测试截图

    4.3 为什么需要将梯度清零（zero_grad()）？

    4.4 在小型合成数据集（如 make_moons）上跑通完整的训练流程，训练脚本 + Loss 下降曲线图

### ===打卡截图要求 请仔细阅读===

微信群截图具体要求："两步走” 与“三要素”

第一步： github 指定issue链接中，上传测试通过截图，作为打卡；

第二步：然后把自己打卡截图发到群里。 微信群截图包括（1）issue 标题 （2） GitHub ID 和（3） 微信群昵称清晰可见

**Comment by Scorpio-Neptune** [comment]
从Tensor到Transformer——构建 GPT 模型 | Task 4 优化器与训练循环
微信：璟翊

4.1 Optimizers 实现思路
Optimizer 的作用是根据反向传播得到的梯度更新模型参数，基本形式为 \(\theta_{t+1}=\theta_t-\eta\nabla L(\theta_t)\)。实现时，Optimizer 基类统一管理需要训练的参数，并提供 zero_grad() 和 step() 接口；SGD 直接沿负梯度方向更新参数，Momentum 利用历史梯度减少震荡，Adam 根据一阶和二阶矩自适应调整不同参数的更新幅度，AdamW 则进一步将 Weight Decay 与梯度更新解耦。测试时主要验证执行 step() 后模型参数是否发生正确变化，以及不同优化器能否正常完成参数更新。

<img width="806" height="438" alt="Image" src="https://github.com/user-attachments/assets/1bbda1cc-8da5-4a4c-973d-aa381ea48fe2" />

4.2 Training 实现思路
Training 将前面实现的各模块组合成完整训练流程。每个 Batch 首先执行 Forward 得到预测结果，然后通过 Loss 计算预测误差，再调用 backward() 获得参数梯度，Optimizer 根据梯度执行 step() 更新参数，之后清除梯度并继续下一个 Batch。不断重复这一过程，使 Loss 逐渐降低、模型参数逐渐逼近更合适的取值。你当前的 Trainer 还进一步支持学习率调度、梯度裁剪、梯度累积以及训练/测试模式切换。

<img width="523" height="407" alt="Image" src="https://github.com/user-attachments/assets/91e39ec6-c973-4d6b-a48a-7dcffba44d07" />

4.3 为什么需要 zero_grad()？
因为反向传播产生的梯度默认会进行累积。如果上一 Batch 的梯度没有清除，那么下一次 backward() 得到的梯度会继续叠加到已有梯度上，使 Optimizer 使用错误的累计梯度更新参数。因此常规训练需要在一次参数更新完成后调用 zero_grad()，将梯度重置，再计算下一批数据的梯度。只有在主动进行 Gradient Accumulation 时，才会故意让多个 Batch 的梯度累积后统一更新；你的 Trainer 也专门支持这种训练方式。

4.4 小型合成数据集完整训练流程 + Loss 下降曲线

<img width="768" height="432" alt="Image" src="https://github.com/user-attachments/assets/97e34e9c-98b6-45f5-812f-5c2f90ed5a88" />
